src/shared/rm-rf.c

```diff
@@ -44,7 +44,7 @@ static int patch_dirfd_mode(
         if (st.st_uid != geteuid())  /* this only works if the UID matches ours */
                 return -EACCES;
 
-        if (fchmod(dfd, (st.st_mode | 0700) & 07777) < 0)
+        if (chmod(FORMAT_PROC_FD_PATH(dfd), (st.st_mode | 0700) & 07777) < 0)
```

Comment:
Sorry, what's the reason of replacing `fchmod` with `chmod`?

Reply:
As `openat_harder()` calls the function for a file descriptor opened with `O_PATH`. `fchmod()` does not work for `O_PATH` file descriptors.

Reply:
use `fchmod_opath()`

Reply:
> It sounds like this chunk should be moved to the commit that adds `openat_harder()` and mentioned in the commit message. The mask change deserves to be its own commit.

Done.

> But, hmm, iti doesn't make much sense to "optimize" by using O_PATH, if we have to go via `/proc` later on. So maybe instead change `openat_harder()` to use a normal fd?

But, /proc is mounted super earlier... And the code is used only in `rm_rf()` at least now. I am not sure we should care such case.
Also, https://github.com/systemd/systemd/pull/26883#discussion_r1143338276.